File: src/main/java/no/nav/pensjon/selvbetjeningopptjening/security/impersonal/ClientCredentialsTokenRequestClient.java
package no.nav.pensjon.selvbetjeningopptjening.security.impersonal;

import no.nav.pensjon.selvbetjeningopptjening.security.oauth2.Oauth2ParamBuilder;
import no.nav.pensjon.selvbetjeningopptjening.security.time.ExpirationChecker;
import no.nav.pensjon.selvbetjeningopptjening.usersession.token.TokenAccessParam;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.stereotype.Component;
import org.springframework.util.MultiValueMap;
import org.springframework.web.reactive.function.client.WebClient;

import static java.util.Objects.requireNonNull;

/**
 * Client for requesting access token based on client credentials grant.
 */
@Component
public class ClientCredentialsTokenRequestClient extends CacheAwareTokenClient {

    private final ClientCredentials credentials;

    public ClientCredentialsTokenRequestClient(@Qualifier("external-call") WebClient webClient,
                                               Oauth2ConfigGetter oauth2ConfigGetter,
                                               ExpirationChecker expirationChecker,
                                               ClientCredentials credentials) {
        super(webClient, oauth2ConfigGetter, expirationChecker);
        this.credentials = requireNonNull(credentials, "credentials");
    }

    @Override
    protected MultiValueMap<String, String> prepareTokenRequestBody(TokenAccessParam accessParam, String unused) {
        return new Oauth2ParamBuilder()
                .tokenAccessParam(accessParam)
                .clientId(credentials.getClientId())
                .clientSecret(credentials.getClientSecret())
                .buildClientCredentialsTokenRequestMap();
    }
}

File: src/main/java/no/nav/pensjon/selvbetjeningopptjening/tech/security/egress/oauth2/clientcred/ClientCredentialsTokenRequestClient.kt
package no.nav.pensjon.selvbetjeningopptjening.tech.security.egress.oauth2.clientcred

import no.nav.pensjon.selvbetjeningopptjening.tech.security.egress.oauth2.OAuth2ParameterBuilder
import no.nav.pensjon.selvbetjeningopptjening.tech.security.egress.token.CacheAwareTokenClient
import no.nav.pensjon.selvbetjeningopptjening.tech.security.egress.token.TokenAccessParameter
import no.nav.pensjon.selvbetjeningopptjening.tech.security.egress.token.validation.ExpirationChecker
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component
import org.springframework.util.MultiValueMap
import org.springframework.web.reactive.function.client.WebClient

@Component
class ClientCredentialsTokenRequestClient(
    @Value("\${azure.openid-config.token-endpoint}") tokenEndpoint: String,
    webClientBuilder: WebClient.Builder,
    expirationChecker: ExpirationChecker,
    private val credentials: ClientCredentials,
    @Value("\${sob.web-client.retry-attempts}") retryAttempts: String
) : CacheAwareTokenClient(
    tokenEndpoint,
    webClientBuilder,
    retryAttempts,
    expirationChecker
) {
    override fun prepareTokenRequestBody(
        accessParameter: TokenAccessParameter,
        audience: String
    ): MultiValueMap<String, String> =
        OAuth2ParameterBuilder()
            .tokenAccessParameter(accessParameter)
            .clientId(credentials.clientId)
            .clientSecret(credentials.clientSecret)
            .buildClientCredentialsTokenRequestMap()
}
